### ISSUE TITLE
Request: NavigationHandler port to webworks

### ISSUE THREAD

**Issue by anphorea** [blackberry/WebWorks-Community-APIs#339]
Currently we can't get the trackpad events directly,
can just pan around or move between elements,
could someone port this NavigationHandler class to webworks api, thank you~

**Comment by timwindsor** [comment]
This one might be very difficult to figure out, but perhaps there's something in the input plugin that we can build on.

**Comment by timwindsor** [comment]
If we follow the structure of the Gamepad plugin, and listen for SCREEN_EVENT_GAMEPAD or SCREEN_EVENT_JOYSTICK events.

Then, this code is not exactly what we need to do, but it's how the Cascades API handles those events, so we need to process it similarly to get the results:

  event.type = ScreenEventTypeTrackPad; 
  event.trackPad.isDisplacementValid = flags & SCREEN_FLAG_DISPLACEMENT_VALID;
  event.trackPad.displacementX = displacement[0];
  event.trackPad.displacementY = displacement[1];
  event.trackPad.pressed = (bool)buttons;
  event.trackPad.modifiers = modifiers;

  event.trackPad.buttonState = buttons;
  event.trackPad.buttonCount = buttonCount;
  for (size_t i = 0; i < 3; ++i) {
    event.trackPad.analog0[i] = analog0[i];
    event.trackPad.analog1[i] = analog1[i];
  }
  event.trackPad.analogCount = analogCount;

**Comment by scheidemanS** [comment]
So if I get the Cordova application window context and create a new window for receiving screen events the same way as in the BarcodeScanner plugin, I can make the new window transparent by setting the window transparency property to SCREEN_TRANSPARENCY_DISCARD. 

Now I can receive all screen events using BPS, however for trackpad movements I do not get  SCREEN_EVENT_JOYSTICK but instead SCREEN_EVENT_POINTER events. 

My current issue is you cannot interact with Cordova application elements (eg. buttons) the way I have it set up.

**Comment by timwindsor** [comment]
I think you want to avoid creating a new screen like BarcodeScanner does. You want to capture events from the current WebView's screen, which is what GamePad does.

The Event type could be wrong. It seems more reasonable to me that they would be Pointer events than Joystick events. Do the actual values line up otherwise?

**Comment by scheidemanS** [comment]
Okay I will look again at how Gamepad does it. 

Yes the values returned match the position on the screen.

**Comment by timwindsor** [comment]
From your note, you are using the BarcodeScanner approach of getting the window and current context. It sounds like when you create a new window, you can get the touch and joystick events, but when you just listen for the current window, you don't get anything - except when the menu appears. Is that correct?

I think this means that the events are getting consumed by the Webview already, so they aren't available to be listened to. This might mean it's not going to work.

If you use the trackpad in a simple app, do you get the mouse cursor on screen? And the same in the Browser?

**Comment by scheidemanS** [comment]
Yes when using the trackpad the mouse cursor appears on the screen for both the simple app and Browser.

**Comment by scheidemanS** [comment]
So what will my next steps be?

**Comment by timwindsor** [comment]
We probably need to try a different tactic.

See this plugin example that handles capacitive keyboard for the touch keyboard on a Passport: https://github.com/blackberry/cordova-blackberry-plugins/blob/master/plugin/com.blackberry.ui.input/src/blackberry10/index.js#L206-L218

I don't think the "UnhandledTouch" event will capture them for us, but we can try these possibilities:
MouseEvent,
GamepadEvent,
UnhandledGamepad.

It shouldn't take too long to try out these events - you can just copy this plugin to start and do some console logging of the events that you can capture.

If it doesn't get the events we need, I think we're stuck on this - so best thing is to write up all the things we tried and post in the issue here, so we can close it but refer back to all the work that was done if needed.

**Comment by scheidemanS** [comment]
Okay so the method in the post above used for receiving touch keyboard events does not work for receiving trackpad events. I tried listening for the 4 events mentioned (MouseEvent, GamepadEvent, UnhandledGamepad, UnhandledTouch), but had no success with any of them. 

Here are the other methods I tried and the results:

First I tried listening for screen events using BPS with a new context , similar to the Gamepad plugin. In this case screen idle events were received; however, no other events were received in the native layer, unless the blackberry side menu was showing. If the side menu was showing then joystick events occurred with coordinates at (0,0,0), when using the trackpad. 

Next I tried creating a new transparent window over top of the application screen the same way as in BarcodeScanner. In index.js, qnx.webplatform.getWebViews() is used to get the handle and group for the current screen. BarcodeScanner sends the screen handle and group to the native layer and uses the handle to get a screen context. Using the context a new window is created and added to the same group as the original window. This window works for receiving screen events using BPS, however it does not allow you to interact with the main application window below it. I also tried using just the context without creating a new window, but this did not work either.
